src/ray/core_worker/test/task_manager_test.cc

```diff
@@ -2542,6 +2577,161 @@ TEST_F(TaskManagerLineageTest, RecoverIntermediateObjectInStreamingGenerator) {
+  TaskSpecification spec;
+  AddInlinedArg(spec, object_id);
+  auto dependencies = ExtractPlasmaDependencies(spec);
+  ASSERT_THAT(dependencies, ::testing::SizeIs(0));
+}
+
+TEST_F(TestExtractPlasmaDependencies, ActorTask) {
+  ObjectID arg1 = ObjectID::FromRandom();
+
+  auto spec = CreateTaskHelper(1, {arg1});
+  spec.GetMutableMessage().set_type(TaskType::ACTOR_TASK);
+
+  ObjectID actor_creation_dummy_id = spec.ActorCreationDummyObjectId();
+
+  auto dependencies = ExtractPlasmaDependencies(spec);
+
+  EXPECT_THAT(dependencies, ::testing::SizeIs(2));
+  EXPECT_THAT(dependencies,
+              ::testing::UnorderedElementsAre(arg1, actor_creation_dummy_id));
+}
+
+TEST_F(TestExtractPlasmaDependencies, MixedArguments) {
+  ObjectID ref_arg = ObjectID::FromRandom();
+  ObjectID nested_ref1 = ObjectID::FromRandom();
+  ObjectID nested_ref2 = ObjectID::FromRandom();
+  ObjectID inlined_arg = ObjectID::FromRandom();
+  ObjectID gpu_object_id = ObjectID::FromRandom();
+
+  auto spec = CreateTaskHelper(1, {ref_arg});
+  AddInlinedArgWithNestedReferences(spec, {nested_ref1, nested_ref2});
+  AddInlinedArg(spec, inlined_arg);
+  AddInlinedArg(spec, gpu_object_id, /*is_gpu_object=*/true);
+
+  auto dependencies = ExtractPlasmaDependencies(spec);
+
+  EXPECT_THAT(dependencies, ::testing::SizeIs(4));
+  EXPECT_THAT(
+      dependencies,
+      ::testing::UnorderedElementsAre(ref_arg, nested_ref1, nested_ref2, gpu_object_id));
+}
+
+TEST_F(TaskManagerTest, TestGPUObjectTaskSuccess) {
+  rpc::Address caller_address;
+  auto spec = CreateTaskHelper(/*num_returns*/ 1,
+                               {},
+                               /*dynamic_returns=*/false,
+                               /*streaming_generator=*/false,
+                               /*generator_backpressure_num_objects*/ -1,
+                               /*enable_tensor_transport=*/true);
+
+  // Add two dependencies, one is an inlined GPU object and the other is
+  // normal inlined object.
+  ObjectID dep1 = ObjectID::FromRandom();
+  AddInlinedArg(spec, dep1, /*is_gpu_object=*/true);
+  ObjectID dep2 = ObjectID::FromRandom();
+  AddInlinedArg(spec, dep2, /*is_gpu_object=*/false);
+  ObjectID dep3 = ObjectID::FromRandom();
+  AddInlinedArg(spec, dep3, /*is_gpu_object=*/false);
+
+  // Check whether the GC callback is registered for the return object.
+  auto return_id = spec.ReturnId(0);
+  EXPECT_CALL(*reference_counter_,
+              AddObjectOutOfScopeOrFreedCallback(return_id, testing::_))
+      .Times(1);
+
+  // Call AddPendingTask to add the task to the task manager.
+  auto object_refs = manager_.AddPendingTask(caller_address, spec, "");
+  ASSERT_EQ(object_refs.size(), 1);
+  ASSERT_EQ(manager_.NumSubmissibleTasks(), 1);
+  ASSERT_EQ(manager_.NumPendingTasks(), 1);
+  ASSERT_TRUE(manager_.IsTaskPending(spec.TaskId()));
+
+  // Only the return object and the actor creation dummy object are in scope.
+  ASSERT_EQ(reference_counter_->NumObjectIDsInScope(), 2);
```

Comment:
Can you also check that the ref counter is updated correctly at different points in the task's lifetime? I.e. when the task's dependencies are resolved and when the task finishes? This was the original issue, if I understand correclty, so we should have a C++ test that captures it

Reply:
I haven’t checked that yet because I think we should only call `ReferenceCounter`'s public functions and mock the virtual public functions defined in `ReferenceCounterInterface` to check the status of reference counter. (Maybe a better pattern is for `ReferenceCounter` to avoid having any public functions that are not defined in `ReferenceCounterInterface`.)

Let me try to figure out how to check it elegantly.